src/web/lib/components/ThemeSaveButton/index.js

```diff
@@ -8,15 +13,29 @@ export const ThemeSaveButton = ({
   storage,
   userHasEdited,
   modifiedSinceSave,
-  setThemeBuilderPanel
+  setThemeBuilderPanel,
+  setUsedStorage,
+  setStorageErrorMessage
 }) => {
   const { themeStorage } = storage;
   const saveTheme = () => {
     if (!modifiedSinceSave) {
       return;
     }
-    themeStorage.put(themeStorage.generateKey(), theme);
-    setThemeBuilderPanel(3);
+
+    try {
+      let space = localStorageSpace();
+      themeStorage.put(themeStorage.generateKey(), theme);
+      setThemeBuilderPanel(3);
+      setUsedStorage({ space });
+    } catch (err) {
+      console.error(err);
+      setStorageErrorMessage(STORAGE_ERROR_MESSAGE);
+      setTimeout(
+        () => setStorageErrorMessage(""),
```

Comment:
Could you move the logic that clears the error to the component that is responsible for rendering the error? And consider clearing the error message if the images were successfully saved.

In the current patch, anything that modifies storage is expected to call `setUsedStorage` afterwards, and apparently also update the UI if needed. Would it be feasible to trigger the UI update from the storage module itself instead, and have consistent handling (e.g. disappearing error message after some timeout)?

There is currently logic to synchronize changes to the storage from other pages (e.g. if the Firefox Color website is opened in two browser windows). You could also consider triggering the notification from that event: https://github.com/mozilla/FirefoxColor/blob/444d50d1dc12fb20b52029000bdb9cb74148692c/src/web/lib/storage.js#L127
(note: this event is only fired for storage changes from other pages. To detect changes by the current page, you need to update the `.put` / `.delete` methods.)

Reply:
So I did move some stuff around. I moved the setLocalStorage to this eventListener function and I moved the setTimeout to the StorageSpaceInformation component (since the storage event listener doesn't get called if exceeds).  The error message should not be set if the images/themes are saved. The error message should only be set if it doesn't go through. I still have the setting part in the catch. in the catch.

I am not sure I am clear on all your suggestions here so please let me know what you think.

Reply:
The "storage" event is only fired if the storage is modified by another page (not the current page). So with your current change, `actions.ui.setUsedStorage` seems to only be dispatched once in practice: When the `lib/storage.js` module is executed.
If you want to automatically propagate storage size changes, then you have to also dispatch `setUsedStorage` whenever the storage is modified by the `put` and `delete` methods in `lib/storage.js`.

> The error message should not be set if the images/themes are saved. The error message should only be set if it doesn't go through.

If the error condition is false (i.e. enough storage), then any previously shown error message should be removed. Hence, I expected the `setStorageErrorMessage` method to be called unconditionally, potentially with an empty string.

>  I still have the setting the error in the catch...

I was thinking of handling storage errors in the `put` method in `lib/storage.js`, so that the UI is always updated regardless of where `put` was called. But wrapping the `put` method in try-catch and calling `setStorageErrorMessage` also works, so it's fine to keep it there.